internal/resolution/entitysources/catalogdsource.go

```diff
@@ -0,0 +1,142 @@
+package entitysources
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/operator-framework/deppy/pkg/deppy"
+	"github.com/operator-framework/deppy/pkg/deppy/input"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/operator-framework/operator-registry/alpha/property"
+
+	catalogd "github.com/operator-framework/catalogd/pkg/apis/core/v1beta1"
+)
+
+// catalogdEntitySource is a source for(/collection of) deppy defined input.Entity, built from content
+// made accessible on-cluster by https://github.com/operator-framework/catalogd.
+// It is an implementation of deppy defined input.EntitySource
+type catalogdEntitySource map[deppy.Identifier]input.Entity
+
+func NewCatalogdEntitySource(ctx context.Context, client client.Client) (catalogdEntitySource, error) {
+	entities := catalogdEntitySource{}
+	bundleMetadatas, packageMetdatas, err := onClusterMetadata(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	for _, bundle := range bundleMetadatas.Items {
+		props := map[string]string{}
+
+		for _, prop := range bundle.Spec.Properties {
+			switch prop.Type {
+			case property.TypePackage:
+				type pkg struct {
+					Name    string `json:"packageName"`
+					Version string `json:"version"`
+				}
+				tmp := pkg{}
+				if err := json.Unmarshal(prop.Value, &tmp); err != nil {
```

Comment:
Mistakes happen? And we should always validate user-derived input?

Reply:
So right now we just throw in whatever we're getting from the [declCfg](https://github.com/operator-framework/catalogd/blob/main/pkg/controllers/core/catalogsource_controller.go#L230) [into the CR](https://github.com/operator-framework/catalogd/blob/main/pkg/controllers/core/catalogsource_controller.go#L257-L260), which ends up looking something like this: 

```yaml
spec:
  catalogSource: catalogsource-sample
  image: quay.io/operatorhubio/argocd-operator@sha256:5f4533de114e9301e836155190dc2c66a969f55085af03a4959a2fb0af74e1f4
  package: argocd-operator
  properties:
  - type: olm.gvk
    value: ewogICAgICAgICAgICAgICAgImdyb3VwIjogImFyZ29wcm9qLmlvIiwKICAgICAgICAgICAgICAgICJraW5kIjogIkFwcFByb2plY3QiLAogICAgICAgICAgICAgICAgInZlcnNpb24iOiAidjFhbHBoYTEiCiAgICAgICAgICAgIH0=
  - type: olm.gvk
    value: ewogICAgICAgICAgICAgICAgImdyb3VwIjogImFyZ29wcm9qLmlvIiwKICAgICAgICAgICAgICAgICJraW5kIjogIkFwcGxpY2F0aW9uIiwKICAgICAgICAgICAgICAgICJ2ZXJzaW9uIjogInYxYWxwaGExIgogICAgICAgICAgICB9
  - type: olm.gvk
    value: ewogICAgICAgICAgICAgICAgImdyb3VwIjogImFyZ29wcm9qLmlvIiwKICAgICAgICAgICAgICAgICJraW5kIjogIkFwcGxpY2F0aW9uU2V0IiwKICAgICAgICAgICAgICAgICJ2ZXJzaW9uIjogInYxYWxwaGExIgogICAgICAgICAgICB9
  - type: olm.gvk
    value: ewogICAgICAgICAgICAgICAgImdyb3VwIjogImFyZ29wcm9qLmlvIiwKICAgICAgICAgICAgICAgICJraW5kIjogIkFyZ29DRCIsCiAgICAgICAgICAgICAgICAidmVyc2lvbiI6ICJ2MWFscGhhMSIKICAgICAgICAgICAgfQ==
  - type: olm.gvk
    value: ewogICAgICAgICAgICAgICAgImdyb3VwIjogImFyZ29wcm9qLmlvIiwKICAgICAgICAgICAgICAgICJraW5kIjogIkFyZ29DREV4cG9ydCIsCiAgICAgICAgICAgICAgICAidmVyc2lvbiI6ICJ2MWFscGhhMSIKICAgICAgICAgICAgfQ==
  - type: olm.package
    value: ewogICAgICAgICAgICAgICAgInBhY2thZ2VOYW1lIjogImFyZ29jZC1vcGVyYXRvciIsCiAgICAgICAgICAgICAgICAidmVyc2lvbiI6ICIwLjUuMCIKICAgICAgICAgICAgfQ==
  relatedImages:
  - image: gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0
    name: ""
  - image: quay.io/argoprojlabs/argocd-operator@sha256:2f5c0d4567607266ccacb91d8c2e3b18c2afe0edaf55855dcb1a06b02173b520
    name: ""
  - image: quay.io/operatorhubio/argocd-operator@sha256:5f4533de114e9301e836155190dc2c66a969f55085af03a4959a2fb0af74e1f4
    name: ""
```  

We'd discussed correcting this and changing the plumbing to make sure we throw in unmarshalled structs there instead, but again, that was superseded with https://github.com/operator-framework/catalogd/issues/38. Hence the tap dance with marshal/unmarshal: just a stop gap so that we're not sweating over details that are going to change soon.

Reply:
@tmshort if it's required, the validation should occur at creation or mutation of the object, not downstream every time the value is used.

Reply:
@anik120 is there an understanding of how long that effort will take? Leaving the software in a more stable spot while the change happens might be useful if it's a long-running effort?

Reply:
It's definitely a "the very next step" for the project. We'd planned it for milestone 3 but it got pushed out in favor of us focusing our effort on the integration for m3, so I'd say milestone 4 is a pretty accurate timeline for when it's about to happen. (We can probably do a v0.2.0 release after m4 and in m5 we can update this integration with the v0.2.0 release, but don't quote me on that yet I'd like to make sure others working in the project are in agreement with "we can do these things we're promising in that timeline" too :) ). But the gist of it is, these are going to change pretty soon.

Reply:
> Mistakes happen? And we should always validate user-derived input?

> @tmshort if it's required, the validation should occur at creation or mutation of the object, not downstream every time the value is used.

All makes sense. I've made a note of it [here](https://github.com/operator-framework/catalogd/issues/38#issuecomment-1522391862) so that we don't miss this

Reply:
I'm not sure I'm understanding the need to unmarshal/marshal. I know that deppy wants maps of properties, not lists, so I would expect some handling to do that conversion. But isn't this just a matter of collecting up the properties that have the same type into `map[string][]json.Message`, and then building the deppy entities from that?